internal/pkg/cli/app_show.go

```diff
@@ -72,109 +62,100 @@ type ShowAppOpts struct {
 }
 
 // Ask asks for fields that are required but not passed in.
-func (opts *ShowAppOpts) Ask() error {
-	// get project
-	if opts.ProjectName() == "" {
-		projectName, err := opts.selectProject()
+func (o *ShowAppOpts) Ask() error {
+	if err := o.askProject(); err != nil {
+		return err
+	}
+	return o.askAppName()
+}
+
+// Validate returns an error if the values provided by the user are invalid.
+func (o *ShowAppOpts) Validate() error {
+	if o.ProjectName() != "" {
+		names, err := o.retrieveProjects()
 		if err != nil {
 			return err
 		}
-		opts.projectName = projectName
+		if !contains(o.ProjectName(), names) {
+			return fmt.Errorf("project '%s' does not exist in the workspace", o.ProjectName())
+		}
 	}
-
-	// get application of the project
-	if opts.appName == "" {
-		appName, err := opts.selectApplication()
+	if o.appName != "" {
+		names, err := o.retrieveApplications()
 		if err != nil {
 			return err
 		}
-		opts.appName = appName
-	}
-
-	return nil
-}
-
-// Validate returns an error if the user inputs are invalid.
-func (opts *ShowAppOpts) Validate() error {
-	proj, err := opts.storeSvc.GetProject(opts.ProjectName())
-	if err != nil {
-		return fmt.Errorf("getting project: %w", err)
+		if !contains(o.appName, names) {
+			return fmt.Errorf("application '%s' does not exist in project '%s'", o.appName, o.ProjectName())
+		}
 	}
-	opts.proj = proj
 
 	return nil
 }
 
 // Execute shows the applications through the prompt.
-func (opts *ShowAppOpts) Execute() error {
-	if opts.appName != "" {
-		if err := opts.retrieveData(); err != nil {
+func (o *ShowAppOpts) Execute() error {
+	if o.appName != "" {
```

Comment:
Is it possible to get to this point with an empty `appName`?

Reply:
Yes, if there is no application in the project. Since we don't want to error out, instead we want to show no applications. If there's no application to list, we don't need users to select and the `appName` will be set to an empty string.

Reply:
So should we simply print a message indicating "no apps found" and return nil, or maybe add a comment indicating that scenario?

Reply:
Yeah I'll print a log message saying "no applications found in project %s".